web/packages/teleterm/src/ui/DocumentAuthorizeWebSession/DocumentAuthorizeWebSession.tsx

```diff
@@ -0,0 +1,195 @@
+
+import Document from 'teleterm/ui/Document';
+import { routing } from 'teleterm/ui/uri';
+import { retryWithRelogin } from 'teleterm/ui/utils';
+import { useAppContext } from 'teleterm/ui/appContextProvider';
+import * as types from 'teleterm/ui/services/workspacesService';
+import { WebSessionRequest } from 'teleterm/ui/services/workspacesService';
+import { useWorkspaceContext } from 'teleterm/ui/Documents';
+
+export function DocumentAuthorizeWebSession(props: {
+  doc: types.DocumentAuthorizeWebSession;
+  visible: boolean;
+}) {
+  const ctx = useAppContext();
+  const { documentsService } = useWorkspaceContext();
+  const rootCluster = ctx.clustersService.findCluster(props.doc.rootClusterUri);
+  const [authorizeAttempt, authorize] = useAsync(async () => {
+    const {
+      response: { confirmationToken },
+    } = await retryWithRelogin(ctx, props.doc.rootClusterUri, () =>
+      ctx.clustersService.authenticateWebDevice(
+        props.doc.rootClusterUri,
+        props.doc.webSessionRequest
+      )
+    );
+    return confirmationToken;
+  });
+  const clusterName = routing.parseClusterName(props.doc.rootClusterUri);
+  const canAuthorize = rootCluster.loggedInUser?.isDeviceTrusted;
+
+  async function authorizeAndCloseDocument() {
+    const [confirmationToken, error] = await authorize();
+    if (!error) {
+      const url = buildAuthorizedSessionUrl(
+        rootCluster,
+        props.doc.webSessionRequest,
+        confirmationToken
+      );
+      // This endpoint verifies the token and "upgrades" the web session and redirects to "/web".
+      window.open(url);
+      closeAndNotify();
+    }
+  }
+
+  function openUnauthorizedAndCloseDocument() {
+    const url = buildUnauthorizedSessionUrl(
+      rootCluster,
+      props.doc.webSessionRequest
+    );
+    window.open(url);
+    closeAndNotify();
+  }
+
+  function closeAndNotify() {
+    documentsService.close(props.doc.uri);
+    ctx.notificationsService.notifyInfo(
+      'Web session has been opened in the browser'
+    );
+  }
+
+  return (
+    <Document visible={props.visible}>
+      <Flex
+        flexDirection="column"
+        maxWidth="680px"
+        width="100%"
+        mx="auto"
+        mt="4"
+        px="5"
+      >
+        <H1 mb="4">Authorize Web Session</H1>
+        <Flex flexDirection="column" gap={3}>
+          {/*It's technically possible to open a deep link to authorize a session on a device that is not enrolled.*/}
+          {!canAuthorize && (
+            <Alert
+              mb={0}
+              details={
+                <>
+                  To authorize a web session, you must first{' '}
+                  <a
+                    href="https://goteleport.com/docs/admin-guides/access-controls/device-trust/guide/#step-22-enroll-device"
+                    target="_blank"
+                  >
+                    enroll your device
+                  </a>
+                  . Then log out of the app, log back in, and try again.
```

Comment:
Forgot to mention it last time, but this should be more specific – to a regular user, "the app" might mean Connect or Web UI. So perhaps "Then log out of Teleport Connect"?

Reply:
Great idea, thanks.